File: JavaMicroserviceBackend/portal-service/src/main/java/com/atesti/portal/domain/model/Ponuda.java
package com.atesti.portal.domain.model;

import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import jakarta.persistence.*;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.Set;

@Entity
@Table(name = "ponude")
@Getter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Ponuda {

    private static final Set<String> VALID_STATUSES = Set.of("nova", "poslana", "odobrena", "odbijena");

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "korisnik_id")
    private Korisnik korisnik;

    @Column(name = "korisnik_id", insertable = false, updatable = false)
    private Long korisnikId;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String opis;

    private String vrstaAtesta;
    private String lokacija;
    private LocalDate zeljeniDatum;

    @Builder.Default
    private String status = "nova";

    @Column(columnDefinition = "TEXT")
    private String odgovor;

    @CreationTimestamp
    private LocalDateTime createdAt;

    @UpdateTimestamp
    private LocalDateTime updatedAt;

    public void updateStatus(String status, String odgovor) {
        if (status != null) {
            if (!VALID_STATUSES.contains(status)) {
                throw new IllegalArgumentException("Nevazeci status: " + status);
            }
            this.status = status;
        }
        if (odgovor != null) {
            this.odgovor = odgovor;
        }
    }
}

File: backend/models/ponuda.js
module.exports = (sequelize, DataTypes) => {
  const Ponuda = sequelize.define('ponuda', {
    id: {
      type: DataTypes.INTEGER,
      primaryKey: true,
      autoIncrement: true
    },
    korisnikId: {
      type: DataTypes.INTEGER,
      allowNull: false
    },
    opis: {
      type: DataTypes.TEXT,
      allowNull: false
    },
    vrstaAtesta: {
      type: DataTypes.STRING,
      allowNull: true
    },
    lokacija: {
      type: DataTypes.STRING,
      allowNull: true
    },
    zeljeniDatum: {
      type: DataTypes.DATEONLY,
      allowNull: true
    },
    status: {
      type: DataTypes.ENUM('nova', 'poslana', 'odobrena', 'odbijena'),
      defaultValue: 'nova'
    },
    odgovor: {
      type: DataTypes.TEXT,
      allowNull: true
    }
  }, {
    timestamps: true,
    tableName: 'ponude'
  });

  Ponuda.associate = (models) => {
    Ponuda.belongsTo(models.korisnik, { foreignKey: 'korisnikId' });
  };

  return Ponuda;
};
